### PR TITLE
feat: informer related behavior on startup and in case of errors

### PR DESCRIPTION
**Pull request by csviri**
Implement behavior requested in #1422 
The sync timeout is also implemented. But might move it to the controller level (as in controller runtime) in a separate PR.

### CODE REVIEW

**Comment by csviri** [comment]
Note that tried to make a distinction of integration tests based on Tag but there were issues with it in surefire plugin

**Review by metacosm — CHANGES_REQUESTED** [review]

File: operator-framework-core/src/main/java/io/javaoperatorsdk/operator/api/config/ConfigurationServiceOverrider.java [hunk 1]
```diff
@@ -27,6 +27,8 @@ public class ConfigurationServiceOverrider {
   private ExecutorService workflowExecutorService;
   private LeaderElectionConfiguration leaderElectionConfiguration;
   private InformerStoppedHandler informerStoppedHandler;
+  private Boolean stopOnInformerErrorDuringStartup;
```
[the remark this answers is not in the document]
Comment (csviri) [reply operator-framework-core/src/main/java/io/javaoperatorsdk/operator/api/config/ConfigurationServiceOverrider.java:30]: yep, it's to check if it is set or not

File: operator-framework-core/src/main/java/io/javaoperatorsdk/operator/processing/event/source/informer/InformerWrapper.java [hunk 2]
```diff
@@ -58,6 +61,24 @@ public void start() throws OperatorException {
...
+      }
+      try {
+        var start = informer.start();
+        // note that in case we don't put here timeout and stopOnInformerErrorDuringStartup is
+        // false, and there is a rbac issue the get never returns; therefore operator never really
+        // starts
+        start.toCompletableFuture().get(configService.cacheSyncTimeout(), TimeUnit.MILLISECONDS);
+      } catch (TimeoutException | ExecutionException e) {
+        log.warn("Informer startup error. Informer: {}", informer, e);
+        if (configService.stopOnInformerErrorDuringStartup()) {
```
[the remark this answers is not in the document]
Comment (csviri) [reply operator-framework-core/src/main/java/io/javaoperatorsdk/operator/processing/event/source/informer/InformerWrapper.java:76]: It retries, with exponential backoff forever.

File: operator-framework-core/src/test/java/io/javaoperatorsdk/operator/processing/event/source/informer/InformerEventSourceTest.java [hunk 3]
```diff
@@ -255,7 +256,8 @@ void informerStoppedHandlerShouldBeCalledWhenInformerStops() {
           MockKubernetesClient.client(Deployment.class, unused -> {
             throw exception;
           }));
-      informerEventSource.start();
+
+      assertThrows(RuntimeException.class, () -> informerEventSource.start());
```
[the remark this answers is not in the document]
Comment (csviri) [reply operator-framework-core/src/test/java/io/javaoperatorsdk/operator/processing/event/source/informer/InformerEventSourceTest.java:260]: added.

File: operator-framework-core/src/main/java/io/javaoperatorsdk/operator/processing/event/source/informer/InformerWrapper.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (metacosm) [reply operator-framework-core/src/main/java/io/javaoperatorsdk/operator/processing/event/source/informer/InformerWrapper.java:76]: Maybe we should log something to say that that's what's going to happen?

[the remark this answers is not in the document]
Comment (csviri) [reply operator-framework-core/src/main/java/io/javaoperatorsdk/operator/processing/event/source/informer/InformerWrapper.java:76]: yep we can too, although fabric also does logging

[the remark this answers is not in the document]
Comment (csviri) [reply operator-framework-core/src/main/java/io/javaoperatorsdk/operator/processing/event/source/informer/InformerWrapper.java:75]: updated

**Review by metacosm — APPROVED** [review]

**Pull request merged**